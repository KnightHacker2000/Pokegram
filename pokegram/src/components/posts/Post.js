import * as React from 'react';
// import Button from '@mui/material/Button';
import Card from '@mui/material/Card';
import CardActions from '@mui/material/CardActions';
import CardContent from '@mui/material/CardContent';
import CardHeader from '@mui/material/CardHeader';
import CardMedia from '@mui/material/CardMedia';
import Container from '@mui/material/Container';
import Grid from '@mui/material/Grid';
import Typography from '@mui/material/Typography';
import Avatar from '@mui/material/Avatar';
import IconButton from '@mui/material/IconButton';
import FavoriteIcon from '@mui/icons-material/Favorite';
import ShareIcon from '@mui/icons-material/Share';
// import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
// import MoreVertIcon from '@mui/icons-material/MoreVert';
import { createTheme, ThemeProvider } from '@mui/material/styles';
import jieni from '../../images/jieni.jpg';

const posts = [{ id: 1 }, { id: 2 }, { id: 3 }, { id: 4 }];

const theme = createTheme();

function Posts() {
  return (
    <ThemeProvider theme={theme}>
      <Container sx={{ py: 8 }} maxWidth="md">
        <Grid container spacing={4}>
          {posts.map((post) => (
            <Grid item key={post.id} xs={12} sm={6} md={4}>
              <Card
                sx={{ height: '100%', display: 'flex', flexDirection: 'column' }}
              >
                <CardHeader
                  avatar={<Avatar />}
                  action={
                    <IconButton aria-label="settings" />
                  }
                  title="Shrimp and Chorizo Paella"
                  subheader="September 14, 2016"
                />
                <CardMedia
                  component="img"
                  height="194"
                  image={jieni}
                  alt="Paella dish"
                />
                <CardContent>
                  <Typography variant="body2" color="text.secondary">
                    {post.id}
                  </Typography>
                </CardContent>
                <CardActions disableSpacing>
                  <IconButton aria-label="add to favorites">
                    <FavoriteIcon />
                  </IconButton>
                  <IconButton aria-label="share">
                    <ShareIcon />
                  </IconButton>
                </CardActions>
              </Card>
            </Grid>
          ))}
        </Grid>
      </Container>
    </ThemeProvider>
  );
}
export default Posts;
