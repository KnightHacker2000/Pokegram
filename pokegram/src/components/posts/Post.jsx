/* eslint-disable max-len */
import PropTypes from 'prop-types';
import * as React from 'react';
import { useState, useEffect, useRef } from 'react';
// import Button from '@mui/material/Button';
import Card from '@mui/material/Card';
import CardActions from '@mui/material/CardActions';
import CardContent from '@mui/material/CardContent';
import CardHeader from '@mui/material/CardHeader';
import CardMedia from '@mui/material/CardMedia';
import Container from '@mui/material/Container';
import DeleteIcon from '@mui/icons-material/Delete';
import Grid from '@mui/material/Grid';
import Typography from '@mui/material/Typography';
import Avatar from '@mui/material/Avatar';
import IconButton from '@mui/material/IconButton';
import FavoriteIcon from '@mui/icons-material/Favorite';
import EditIcon from '@mui/icons-material/Edit';
import AddCommentIcon from '@mui/icons-material/AddComment';
import AccountCircleIcon from '@mui/icons-material/AccountCircle';
import { createTheme, ThemeProvider } from '@mui/material/styles';
import postsService from '../../services/postsService';
import userService from '../../services/userService';
// import pokemon from '../../images/pikachu.jpg';
import Comment from './Comment';
import TagPhoto from './Tag';
import HomeState from '../../models/homeState';
import Edit from '../update_post/update_post';

const theme = createTheme();

function Posts(props) {
  const { homeStates } = props;
  // console.log(homeStates);
  // console.log(homeStates.myUID);
  // console.log(homeStates.UID);
  const [postList, setPostList] = useState([{
    id: 1,
    username: 'Rachel',
    timestamp: 'Wed Dec 01 2021 03:24:00 GMT-0500 (Eastern Standard Time)',
    type: 'photo',
    content_url: 'http://img4.wikia.nocookie.net/__cb20140328190757/pokemon/images/thumb/2/21/001Bulbasaur.png/200px-001Bulbasaur.png',
    numLike: 10,
    description: 'pikapika!',
    commentRefs: [
      1234,
      5678,
      9101
    ],
    users: [
      1234,
      5678,
      9101
    ]
  }]);
  // const [postList, setPostList] = useState([]);
  const [renderEdit, setrenderEdit] = useState(false);
  const [renderComment, setrenderComment] = useState(false);
  const [renderTagging, setrenderTagging] = useState(false);
  // const [isLike, setLike] = useState(false);
  const [likePosts, setlikePosts] = useState([]);
  const [, updateState] = React.useState();
  const [editPostId, setEditPostId] = useState(-1);
  const [commentPostId, setcommentPostId] = useState(-1);
  const [tagPostId, settagPostId] = useState(-1);
  // const [deletePostId, setdeletePostId] = useState(-1);
  const firstRendering = useRef(true);
  const canEdit = homeStates.UID === homeStates.myUID;
  const forceUpdate = React.useCallback(() => updateState({}), []);

  useEffect(() => {
    // const params = '{"userId": 1}';
    async function fetchallpostsData() {
      const data = await postsService.getAllPosts();
      // console.log(data);
      setPostList(data);
    }

    async function fetchpostsbyusername(userName) {
      const data = await postsService.getPostsByUserName(userName);
      setPostList(data);
    }

    async function getUser() {
      const userparams = { userId: homeStates.myUID };
      const user = await userService.getUserById((userparams));
      setlikePosts(user.likedPosts);
    }

    if (firstRendering.current) {
      firstRendering.current = false;
      if (homeStates.UID === -1) {
        fetchallpostsData();
      } else {
        console.log('entered personal!');
        fetchpostsbyusername(homeStates.myUID);
      }
      getUser();
      // putData();
    }
  });

  function rendermedia(post) {
    let ret;
    if (post.type === 'video') {
      ret = (
        <CardMedia className="media" component="video" height={200} width={400} src={post.content_url} alt="post video" autoPlay controls />
      );
    } else {
      ret = (
        <CardMedia className="media" component="img" height={200} width={400} src={post.content_url} alt="post image" />
      );
    }
    return ret;
  }

  const handleCardClick = (event) => {
    event.preventDefault();
    setEditPostId(event.currentTarget.getAttribute('data-index'));
    setrenderEdit(true);
    forceUpdate();
  };

  const handleEdit = () => {
    setrenderEdit(false);
    firstRendering.current = true;
    forceUpdate();
  };

  const handleAddCommentClick = (event) => {
    event.preventDefault();
    setcommentPostId(event.currentTarget.getAttribute('data-index'));
    setrenderComment(true);
    forceUpdate();
  };

  // const handleLikeClick = (event) => {
  // event.preventDefault();
  // setlikePostId(event.currentTarget.getAttribute('data-index'));
  // setrenderLike(true);
  // forceUpdate();
  // };

  const handlePhotoTagClick = (event) => {
    event.preventDefault();
    settagPostId(event.currentTarget.getAttribute('data-index'));
    setrenderTagging(true);
    forceUpdate();
  };

  const handleComment = () => {
    setrenderComment(false);
    firstRendering.current = true;
    forceUpdate();
  };

  const handleTagPost = () => {
    setrenderTagging(false);
    firstRendering.current = true;
    forceUpdate();
  };

  const handleLikePost = async (event) => {
    // console.log(event.currentTarget.getAttribute('data-index'));
    // console.log(isLike);
    const postId = event.currentTarget.getAttribute('data-index');
    // const numpostId = parseInt(postId, 10);
    // const params = `{"postId":${postId} }`;
    const userparams = { userId: homeStates.myUID };
    const user = await userService.getUserById(userparams);
    const post = await postsService.getPostsById(postId);
    await userService.addlike(user, postId);
    post.numLike += 1;
    await postsService.updatePost(post);
    likePosts.push(postId);
    setlikePosts(likePosts);
    firstRendering.current = true;
    forceUpdate();
  };
  const handleUnLikePost = async (event) => {
    const postId = event.currentTarget.getAttribute('data-index');
    // const numpostId = parseInt(postId, 10);
    // const params = `{"postId":${postId} }`;
    const userparams = { userId: homeStates.myUID };
    const user = await userService.getUserById((userparams));
    const post = await postsService.getPostsById(postId);
    const pIndex = likePosts.indexOf(postId);
    likePosts.splice(pIndex, 1);
    await userService.removeLike(user, postId);
    post.numLike -= 1;
    await postsService.updatePost(post);
    setlikePosts(likePosts);
    firstRendering.current = true;
    forceUpdate();
  };

  const handleDeletePost = async (event) => {
    // console.log(event.currentTarget.getAttribute('data-index'));
    const postId = (event.currentTarget.getAttribute('data-index'));
    // setdeletePostId(postId);
    // console.log(postId);
    // console.log(deletePostId);
    await postsService.deletePost(postId);
    firstRendering.current = true;
    forceUpdate();
  };

  const handleAvatarClick = (event) => {
    // homeStates.handleHomeStates(false, false, false, false, true, homeStates.UID);
    event.preventDefault();
    const otherUID = event.currentTarget.getAttribute('data-index');
    homeStates.handleHomeStates(false, true, false, false, false, otherUID);
  };

  function likebuton(postId) {
    let ret;
    if (likePosts && likePosts.includes(postId)) {
      ret = (
        <IconButton
          sx={{ color: 'red' }}
          aria-label="add to favorites"
          onClick={handleUnLikePost}
          data-index={postId}
        >
          <FavoriteIcon />
        </IconButton>
      );
    } else {
      ret = (
        <IconButton
          aria-label="add to favorites"
          onClick={handleLikePost}
          data-index={postId}
        >
          <FavoriteIcon />
        </IconButton>
      );
    }
    return ret;
  }

  return (
    <ThemeProvider theme={theme}>
      {renderEdit && <Edit pid={editPostId} handleEditState={handleEdit} />}
      {renderComment && <Comment uid={homeStates.myUID} pid={commentPostId} handleCommentState={handleComment} />}
      {renderTagging && <TagPhoto pid={tagPostId} handleTagState={handleTagPost} />}
      <Container sx={{ py: 8 }} maxWidth="md">
        <Grid container spacing={4}>
          {postList.map((post) => (
            <Grid item key={post.id} xs={12} sm={6} md={4}>
              <Card
                sx={{ height: '100%', display: 'flex', flexDirection: 'column' }}
              >
                <CardHeader
                  avatar={<Avatar alt="pikachu" sx={{ m: 1 }} src="http://img4.wikia.nocookie.net/__cb20140328190757/pokemon/images/thumb/2/21/001Bulbasaur.png/200px-001Bulbasaur.png" onClick={handleAvatarClick} data-index={post.username} />}
                  action={
                    <IconButton aria-label="settings" />
                  }
                  title={post.username}
                  subheader={post.timestamp.toString()}
                />
                {rendermedia(post)}
                <CardContent>
                  <Typography variant="body2" color="text.secondary" maxWidth="20vw">
                    {post.description}
                  </Typography>
                </CardContent>
                <CardActions disableSpacing>
                  {likebuton(post.id)}
                  { canEdit && (
                    <IconButton
                      aria-label="edit"
                      onClick={handleCardClick}
                      data-index={post.id}
                    >
                      <EditIcon />
                    </IconButton>
                  )}
                  <IconButton
                    aria-label="comment"
                    onClick={handleAddCommentClick}
                    data-index={post.id}
                    data-testid="test_addComm"
                  >
                    <AddCommentIcon />
                  </IconButton>
                  <IconButton
                    aria-label="tagging"
                    onClick={handlePhotoTagClick}
                    data-index={post.id}
                  >
                    <AccountCircleIcon />
                  </IconButton>
                  { canEdit && (
                    <IconButton
                      aria-label="delete"
                      onClick={handleDeletePost}
                      data-index={post.id}
                    >
                      <DeleteIcon />
                    </IconButton>
                  )}
                </CardActions>
              </Card>
            </Grid>
          ))}
        </Grid>
      </Container>
    </ThemeProvider>
  );
}
Posts.propTypes = {
  homeStates: PropTypes.instanceOf(HomeState)
};

Posts.defaultProps = {
  homeStates: null
};
export default Posts;
