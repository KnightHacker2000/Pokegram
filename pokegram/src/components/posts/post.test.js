/**
 * @jest-environment jsdom
 */
import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Post from './Post';
import HomeState from '../../models/homeState';
import '@testing-library/jest-dom/extend-expect';

const parentStates = new HomeState(
  2,
  (isLogOut, newIsProf, newIsAct, newIsUp, newIsPosts, newUID) => {
    console.log(isLogOut, newIsProf, newIsAct, newIsUp, newIsPosts, newUID);
  },
  false,
  false,
  true,
  false,
  2
);

test('test post snapshot', () => {
  const renderer = new ShallowRenderer();
  renderer.render(<Post homeStates={parentStates} />);
  const postComp = renderer.getRenderOutput();
  expect(postComp).toMatchSnapshot();
});

test('test add fav', () => {
  const renderer = new ShallowRenderer();
  renderer.render(<Post homeStates={parentStates} />);
  const postComp = renderer.getRenderOutput();
  postComp.props.children[3].props.children.props.children[0].props.children.props.children[3].props.children[0].props.onClick({currentTarget: {getAttribute: ()=>{}},preventDefault: () => {}});
});

test('test button', () => {
  const renderer = new ShallowRenderer();
  renderer.render(<Post homeStates={parentStates} />);
  const postComp = renderer.getRenderOutput();
  postComp.props.children[3].props.children.props.children[0].props.children.props.children[3].props.children[1].props.onClick({currentTarget: {getAttribute: ()=>{}},preventDefault: () => {}});
});

test('test button', () => {
  const renderer = new ShallowRenderer();
  renderer.render(<Post homeStates={parentStates} />);
  const postComp = renderer.getRenderOutput();
  postComp.props.children[3].props.children.props.children[0].props.children.props.children[3].props.children[2].props.onClick({currentTarget: {getAttribute: ()=>{}},preventDefault: () => {}});
});

test('test button', () => {
  const renderer = new ShallowRenderer();
  renderer.render(<Post homeStates={parentStates} />);
  const postComp = renderer.getRenderOutput();
  postComp.props.children[3].props.children.props.children[0].props.children.props.children[3].props.children[3].props.onClick({currentTarget: {getAttribute: ()=>{}},preventDefault: () => {}});
});

test('test button', () => {
  const renderer = new ShallowRenderer();
  renderer.render(<Post homeStates={parentStates} />);
  const postComp = renderer.getRenderOutput();
  postComp.props.children[3].props.children.props.children[0].props.children.props.children[3].props.children[4].props.onClick({currentTarget: {getAttribute: ()=>{}},preventDefault: () => {}});
});