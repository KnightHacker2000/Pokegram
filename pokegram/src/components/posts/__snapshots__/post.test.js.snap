// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Post snapshot 1`] = `
<div
  className="MuiContainer-root MuiContainer-maxWidthMd css-jcfe1v-MuiContainer-root"
>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4 css-480o17-MuiGrid-root"
  />
</div>
`;
